### PR TITLE
Fix ability to disable IPv6 in datapath

### PR DESCRIPTION
**Pull request by tgraf**
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/6490)

### CODE REVIEW

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-me-please

**Review by joestringer — APPROVED** [review]
LGTM.

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
The runtime tests are exposing a bug in the datapath:
```
Dec 20 21:40:11 runtime cilium-agent[9228]: level=info msg="Serving cilium health at http://[::]:4240" subsys=health-server
Dec 20 21:40:11 runtime cilium-agent[9228]: level=error msg="Failed to compile bpf_lxc.o: exit status 1" compiler-pid=9818 linker-pid=9820 subsys=datapath-loader
Dec 20 21:40:11 runtime cilium-agent[9228]: level=warning msg="In file included from /var/lib/cilium/bpf/bpf_lxc.c:39:" subsys=datapath-loader
Dec 20 21:40:11 runtime cilium-agent[9228]: level=warning msg="In file included from /var/lib/cilium/bpf/lib/l3.h:24:" subsys=datapath-loader
Dec 20 21:40:11 runtime cilium-agent[9228]: level=warning msg="/var/lib/cilium/bpf/lib/eps.h:124:59: error: use of undeclared identifier 'IPCACHE6_PREFIXES'" subsys=datapath-loader
Dec 20 21:40:11 runtime cilium-agent[9228]: level=warning msg="LPM_LOOKUP_FN(lookup_ip6_remote_endpoint, union v6addr *, IPCACHE6_PREFIXES," subsys=datapath-loader
Dec 20 21:40:11 runtime cilium-agent[9228]: level=warning msg="                                                          ^" subsys=datapath-loader
Dec 20 21:40:11 runtime cilium-agent[9228]: level=warning msg="/var/lib/cilium/bpf/lib/eps.h:124:1: error: invalid application of 'sizeof' to an incomplete type 'int []'" subsys=datapath-loader
Dec 20 21:40:11 runtime cilium-agent[9228]: level=warning msg="LPM_LOOKUP_FN(lookup_ip6_remote_endpoint, union v6addr *, IPCACHE6_PREFIXES," subsys=datapath-loader
Dec 20 21:40:11 runtime cilium-agent[9228]: level=warning msg="^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~" subsys=datapath-loader
Dec 20 21:40:11 runtime cilium-agent[9228]: level=warning msg="/var/lib/cilium/bpf/lib/eps.h:111:26: note: expanded from macro 'LPM_LOOKUP_FN'" subsys=datapath-loader
Dec 20 21:40:11 runtime cilium-agent[9228]: level=warning msg="        const int size = (sizeof(prefixes) / sizeof(prefixes[0]));      \\" subsys=datapath-loader
Dec 20 21:40:11 runtime cilium-agent[9228]: level=warning msg="                                ^~~~~~~~~~" subsys=datapath-loader
Dec 20 21:40:11 runtime cilium-agent[9228]: level=warning msg="/var/lib/cilium/bpf/lib/eps.h:126:51: error: use of undeclared identifier 'IPCACHE4_PREFIXES'" subsys=datapath-loader
Dec 20 21:40:11 runtime cilium-agent[9228]: level=warning msg="LPM_LOOKUP_FN(lookup_ip4_remote_endpoint, __be32, IPCACHE4_PREFIXES," subsys=datapath-loader
Dec 20 21:40:11 runtime cilium-agent[9228]: level=warning msg="                                                  ^" subsys=datapath-loader
Dec 20 21:40:11 runtime cilium-agent[9228]: level=warning msg="/var/lib/cilium/bpf/lib/eps.h:126:1: error: invalid application of 'sizeof' to an incomplete type 'int []'" subsys=datapath-loader
Dec 20 21:40:11 runtime cilium-agent[9228]: level=warning msg="LPM_LOOKUP_FN(lookup_ip4_remote_endpoint, __be32, IPCACHE4_PREFIXES," subsys=datapath-loader
Dec 20 21:40:11 runtime cilium-agent[9228]: level=warning msg="^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~" subsys=datapath-loader
Dec 20 21:40:11 runtime cilium-agent[9228]: level=warning msg="/var/lib/cilium/bpf/lib/eps.h:111:26: note: expanded from macro 'LPM_LOOKUP_FN'" subsys=datapath-loader
Dec 20 21:40:11 runtime cilium-agent[9228]: level=warning msg="        const int size = (sizeof(prefixes) / sizeof(prefixes[0]));      \\" subsys=datapath-loader
Dec 20 21:40:11 runtime cilium-agent[9228]: level=warning msg="                                ^~~~~~~~~~" subsys=datapath-loader
Dec 20 21:40:11 runtime cilium-agent[9228]: level=warning msg="4 errors generated." subsys=datapath-loader
```

**Comment by tgraf** [comment]
The runtime test breakage is caused by the removal of OpenAfterMount(). Removing that commit and limiting this PR to fixing the disable IPv6 feature in the datapath.

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
@joestringer I've changed the way `bpf/Makefile` handles the compilation of various option combinations and added various combinations. Please have a look.

**Review by joestringer — CHANGES_REQUESTED** [review]
This is exactly the sort of thing I was thinking about, fixes #2906.

However, this again breaks the standard compilation options which are used to test complexity. On line 61 (and other similar lines), we compile the final version of the program which will be used to test complexity in, eg, the ginkgo verifier tests. With the current state of this PR, that will compile the BPF programs with no support for IPv4 or IPv6. For that compilation, we want to enable as much complexity as possible. This is why I've been trying to push the makefile and bpf code to use "disable XXX"-style #defines, so that the default options if nothing is specified will have the maximum complexity. Perhaps a more natural alternative would be to define an `ALL_OPTIONS` or `MAX_COMPLEXITY` option in the Makefile which will be used on lines like 61 to ensure that the most complex program will be compiled and left on the filesystem for later tests to consume.

**Comment by tgraf** [comment]
> This is why I've been trying to push the makefile and bpf code to use "disable XXX"-style #defines, so that the default options if nothing is specified will have the maximum complexity. Perhaps a more natural alternative would be to define an `ALL_OPTIONS` or `MAX_COMPLEXITY` option in the Makefile which will be used on lines like 61 to ensure that the most complex program will be compiled and left on the filesystem for later tests to consume.

Then why do we have all the combinations before if only the last invocation triggers the complexity barrier? I'm a bit lost now.

I absolutely dislike having a mix of both ENABLE_IPV6 *and* DISABLE_IPV6.

**Comment by joestringer** [comment]
> Then why do we have all the combinations before if only the last invocation triggers the complexity barrier? I'm a bit lost now.

The earlier combinations test that those option combinations compile, which I think is still a useful check. If we also test that the most complex set of options can load into any kernel, then we get a fairly strong guarantee that we can compile and load the datapath with any set of options. That's why we leave the most complex version on the filesystem for usage in the ginkgo tests.

> I absolutely dislike having a mix of both ENABLE_IPV6 and DISABLE_IPV6.

I agree, this is ugly, I prefer the option you've posted here---we just need to ensure that the most complex version is the one that is left in the directory.

**Comment by tgraf** [comment]
> The earlier combinations test that those option combinations compile, which I think is still a useful check. If we also test that the most complex set of options can load into any kernel, then we get a fairly strong guarantee that we can compile and load the datapath with any set of options. That's why we leave the most complex version on the filesystem for usage in the ginkgo tests.

OK. It would have been great to document this a bit. It was not obvious to me at all.

I've fixed it up and also fixed the LB tests

**Review by joestringer — APPROVED** [review]
LGTM, thanks.

**Comment by joestringer** [comment]
> OK. It would have been great to document this a bit. It was not obvious to me at all.

Anywhere you'd prefer this to be? Inline in the Makefile or in the docs or the ginkgo test?

**Comment by tgraf** [comment]
@joestringer I've added comments to the Makefile now. It's at least clear to me now ;-)

**Comment by tgraf** [comment]
test-me-please

```
github.com/cilium/cilium/pkg/envoy.RequiredEnvoyVersionSHA=84ee839e1d78ef858a39e390288ad417d35bb1d4"' -tags lockdebug -o cilium-agent
00:31:07     k8s1-1.13: Sending build context to Docker daemon  102.4MB
00:31:07     k8s1-1.13: Step 1/18 : FROM quay.io/cilium/cilium-envoy:84ee839e1d78ef858a39e390288ad417d35bb1d4 as cilium-envoy
00:31:07     k8s1-1.13: Get https://quay.io/v2/: EOF
00:31:07     k8s1-1.13: Makefile:149: recipe for target 'docker-image-no-clean' failed
00:31:07     k8s1-1.13: make: *** [docker-image-no-clean] Error 1
```

File: bpf/Makefile [hunk 1]
```diff
@@ -19,46 +19,67 @@ $(BPF_SIMPLE): %.o: %.ll
...
 
 bpf_lxc.ll: bpf_lxc.c $(LIB)
 	$(QUIET) set -e; \
 	$(foreach OPTS,$(LXC_OPTIONS), \
-		$(ECHO_CC) " [$(OPTS)]"; \
-		${CLANG} ${OPTS} ${CLANG_FLAGS} -c $< -o $@; \
+		$(ECHO_CC) " [$(subst :,$(space),$(OPTS))]"; \
+		${CLANG} $(subst :,$(space),$(OPTS)) ${CLANG_FLAGS} -c $< -o $@; \
 		${LLC} ${LLC_FLAGS} -o /dev/null $@; )
 	@$(ECHO_CC)
```
Comment (joestringer) [inline bpf/Makefile:81]: Might be nice to get the annotations here like we do for the compile options, ie `@$(ECHO_CC) " [${MAX_LXC_OPTONS}]"` (and in the other spot), but not a biggie.

Comment (tgraf) [reply bpf/Makefile:81]: The entire command line is printed:
```
clang "-DENABLE_IPV4 -DENABLE_IPV6" -I/home/vagrant/go/src/github.com/cilium/cilium/bpf/include -I/home/vagrant/go/src/github.com/cilium/cilium/bpf -D__NR_CPUS__=2 -O2 -g -target bpf -emit-llvm -Wall -Werror -Wno-address-of-packed-member -Wno-unknown-warning-option -c bpf_lxc.c -o bpf_lxc.ll
```

I don't actually understand the logic of the existing annotations. It seems that all annotations are printed first and then all compilations happen. I guess it's useful to see what combinations are tested but it doesn't actually allow to correlate a compile error to a combination.

**Comment by tgraf** [comment]
test-me-please

File: bpf/Makefile (hunk 1 above)
Comment (joestringer) [reply bpf/Makefile:81]: @tgraf Here's how it works for me (I shortcut all the make parameters with aliases though):

```
$ git di                                                                                                                                     
diff --git a/bpf/lib/common.h b/bpf/lib/common.h                                                                                                                                         
index 5a31d0dc784d..659ddbe02096 100644                                                                                                                                                  
--- a/bpf/lib/common.h                                                                                                                                                                   
+++ b/bpf/lib/common.h                                                                                                                                                                   
@@ -60,6 +60,10 @@                                                                                                                                                                       
 #define CILIUM_CALL_IPV6_TO_LXC                        12                                                                                                                               
 #define CILIUM_CALL_SIZE                       13                                                                                                                                       
                                                                                                                                                                                         
+#if defined(ENABLE_IPV4) && defined(ENABLE_IPV6)                                                                                                                                        
+invalid-c                                                                                                                                                                               
+#endif                                                                                                                                                                                  
+                                                                                                                                                                                        
 typedef __u64 mac_t;                                                                                                                                                                    
                                                                                                                                                                                         
 union v6addr {                                                                                                                                                                          
$ make --quiet V=0 -j1 -C bpf                                                                                                                
  CC    bpf/bpf_lxc.ll  [-DSKIP_DEBUG]                                                                                                                                                   
  CC    bpf/bpf_lxc.ll  [-DHAVE_LPM_MAP_TYPE]                                                                                                                                            
  CC    bpf/bpf_lxc.ll  [-DHAVE_LRU_MAP_TYPE]                                                                                                                                            
  CC    bpf/bpf_lxc.ll  [-DENABLE_IPV4]                                                                                                                                                  
  CC    bpf/bpf_lxc.ll  [-DENABLE_IPV4 -DHAVE_LPM_MAP_TYPE]                                                                                                                              
  CC    bpf/bpf_lxc.ll  [-DENABLE_IPV4 -DHAVE_LPM_MAP_TYPE -DHAVE_LRU_MAP_TYPE]                                                                                                          
  CC    bpf/bpf_lxc.ll  [-DENABLE_IPV6]                                                                                                                                                  
  CC    bpf/bpf_lxc.ll  [-DENABLE_IPV6 -DHAVE_LPM_MAP_TYPE]
  CC    bpf/bpf_lxc.ll  [-DENABLE_IPV6 -DHAVE_LPM_MAP_TYPE -DHAVE_LRU_MAP_TYPE]
  CC    bpf/bpf_lxc.ll  [-DENABLE_IPV6 -DENABLE_IPV4]
In file included from bpf_lxc.c:31:
/home/joe/work/src/github.com/cilium/cilium/bpf/lib/common.h:64:1: error: unknown type name 'invalid'                                                                                   
invalid-c
^
/home/joe/work/src/github.com/cilium/cilium/bpf/lib/common.h:64:8: error: expected identifier or '('                                                                                    
invalid-c
       ^
/home/joe/work/src/github.com/cilium/cilium/bpf/lib/common.h:178:2: error: unknown type name 'mac_t'                                                                                    
        mac_t           mac;
        ^
/home/joe/work/src/github.com/cilium/cilium/bpf/lib/common.h:179:2: error: unknown type name 'mac_t'                                                                                    
        mac_t           node_mac;
        ^
/home/joe/work/src/github.com/cilium/cilium/bpf/lib/common.h:526:2: error: unknown type name 'mac_t'                                                                                    
        mac_t node_mac;
        ^
...
```

**Pull request merged**